hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -2430,6 +2432,71 @@ private class Emit(
           }
         }
 
+      case x@NDArraySlice(child, slicesIR) =>
+        val childEmitter = deforest(child)
+
+        val slicest = emit(slicesIR, env, resultRegion, None)
+        val slicesValueAddress = mb.newField[Long]
+        val slicesTuple = new CodePTuple(coerce[PTuple](slicesIR.pType), region, slicesValueAddress)
+
+        val codeSlices = ArrayBuffer[(Code[Long], Code[Long], Code[Long])]()
+        val codeRefMap = mutable.Map[Int, Code[Long]]()
+
+        var missingSliceElements = const(false)
+
+        coerce[PTuple](slicesIR.pType).types.zipWithIndex.foreach { case (sliceOrIndex, dim) =>
+          sliceOrIndex match {
+            case p: PTuple => {
+              val oneSlice = new CodePTuple(p, region, slicesTuple(dim))
+              missingSliceElements = missingSliceElements || oneSlice.isMissing(0) || oneSlice.isMissing(1) || oneSlice.isMissing(2)
+              codeSlices += ((oneSlice[Long](0), oneSlice[Long](1), oneSlice[Long](2)))
+            }
+            case _: PInt64 => {
+              codeRefMap += dim -> slicesTuple(dim)
+            }
+          }
+        }
+
+        val outputShape = codeSlices.toArray.map { case (start, stop, step) =>
+          (step >= 0L && start <= stop).mux(
+            const(1L) + ((stop - start) - const(1L)) / step,
+            (step < 0L && stop <= start).mux(
+              (((stop - start) + const(1L)) / step) + const(1L),
```

Comment:
We should keep the order consistent between these two branches for all of our sanity.